File: app/src/main/java/ceui/lisa/adapters/IAdapterWithHeadView.java
package ceui.lisa.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import java.util.List;

import ceui.lisa.databinding.RecyRecmdHeaderBinding;
import ceui.lisa.models.IllustsBean;

public class IAdapterWithHeadView extends IAdapter {

    private IllustHeader mIllustHeader = null;
    private String type = "";

    public IAdapterWithHeadView(List<IllustsBean> targetList, Context context, String type) {
        super(targetList, context);
        this.type = type;
    }

    @Override
    public int headerSize() {
        return 1;
    }

    @Override
    public ViewHolder<RecyRecmdHeaderBinding> getHeader(ViewGroup parent) {
        mIllustHeader = new IllustHeader(
            RecyRecmdHeaderBinding.inflate(
                        LayoutInflater.from(mContext),
                        null,
                        false
                ), type
        );
        mIllustHeader.initView(mContext);
        return mIllustHeader;
    }

    public void setHeadData(List<IllustsBean> illustsBeans) {
        if (mIllustHeader != null) {
            mIllustHeader.show(mContext, illustsBeans);
        }
    }

    @Override
    public void onViewAttachedToWindow(@NonNull RecyclerView.ViewHolder holder) {
        ViewGroup.LayoutParams lp = holder.itemView.getLayoutParams();
        if (lp instanceof StaggeredGridLayoutManager.LayoutParams
                && holder.getLayoutPosition() == 0) {
            StaggeredGridLayoutManager.LayoutParams p = (StaggeredGridLayoutManager.LayoutParams) lp;
            p.setFullSpan(true);
        }
    }
}

File: app/src/main/java/ceui/lisa/adapters/IAdapterWithHeadView.kt
package ceui.lisa.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import ceui.lisa.databinding.RecyRecmdHeaderBinding
import ceui.lisa.models.IllustsBean

class IAdapterWithHeadView(
    targetList: List<IllustsBean>,
    context: Context,
    private val type: String
) : IAdapter(targetList, context) {

    private var mIllustHeader: IllustHeader? = null

    override fun headerSize(): Int = 1

    override fun getHeader(parent: ViewGroup): ViewHolder<RecyRecmdHeaderBinding> {
        val header = IllustHeader(
            RecyRecmdHeaderBinding.inflate(
                LayoutInflater.from(mContext),
                null,
                false
            ),
            type
        )
        header.initView(mContext)
        mIllustHeader = header
        return header
    }

    fun setHeadData(illustsBeans: List<IllustsBean>) {
        mIllustHeader?.show(mContext, illustsBeans)
    }

    override fun onViewAttachedToWindow(holder: RecyclerView.ViewHolder) {
        val lp = holder.itemView.layoutParams
        if (lp is StaggeredGridLayoutManager.LayoutParams && holder.layoutPosition == 0) {
            lp.isFullSpan = true
        }
    }
}
